babel/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -1583,3 +1583,49 @@ SqlNode SqlSelectTopN(SqlParserPos pos) :
             SqlLiteral.createBoolean(withTies, pos));
     }
 }
+
+SqlNode SqlHexCharStringLiteral() :
+{
+    final NlsString p;
+    final String formatString;
+    final String charSet;
+    CharacterSet charSetEnum = null;
+    final HexCharLiteralFormat format;
+}
+{
+    (
+        <PREFIXED_HEX_STRING_LITERAL>
+        {
+            charSet = SqlParserUtil
+                      .trim(SqlParserUtil.getCharacterSet(token.image), " ")
+                      .toUpperCase();
+            if (charSet.equals("LATIN")) {
+                charSetEnum = CharacterSet.LATIN;
+            } else if (charSet.equals("UNICODE")) {
+                charSetEnum = CharacterSet.UNICODE;
+            } else if (charSet.equals("GRAPHIC")) {
+               charSetEnum = CharacterSet.GRAPHIC;
+            } else if (charSet.equals("KANJISJIS")) {
+              charSetEnum = CharacterSet.KANJISJIS;
+            } else {
+              throw SqlUtil.newContextException(getPos(),
+              RESOURCE.unknownCharacterSet(charSet));
+            }
+        }
+    |
+        <QUOTED_HEX_STRING>
+    )
+    {
+        String[] tokens = token.image.split("'");
+        p = new NlsString(tokens[1], null, null);
```

Comment:
Do we need to check that the length of tokens is always at least 2?

Reply:
Since the parser would already match either <PREFIXED_HEX_STRING_LITERAL> or <QUOTED_HEX_STRING> at this point, the token is guarantee to be in the format [_<characterSet>]'<hexValue>'[XC|XCV|XCF]. Therefore, we have guarantee that these string manipulation won't fail.

Reply:
Can you add a comment here about that?

Reply:
Sure, will add a piece of comment to clarify.